### PR TITLE
Rich markdown editor

### LINKED ISSUE #281: Markdown editor
Currently writing markdown is very barebones: you can write text and preview it. This issue is about creating a more rich experience by making an interactive markdown editor for comments/posts.

- [x] Bold, italic, headings are insertable with a button
- [x] Links are insertable with a button
- [x] Images are insertable with a button
- [x] https://github.com/krawieck/lemmur/issues/216
- [x] Users can be tagged using an autocomplete

### PR DESCRIPTION
**Pull request by krawieck**
https://user-images.githubusercontent.com/10037914/186002239-f76d6b43-788b-4037-af6d-c4603d680f33.mov

closes #281

- [x] bold
- [x] italics
- [x] link
- [x] image
- [x] user tagging
- [x] instance tagging
- [x] header (6 levels!)
- [x] strikethrough
- [x] quote
- [x] list
- [x] code
- [x] superscript
- [x] subscript
- [x] spoiler

### CODE REVIEW

File: lib/widgets/editor/editor_toolbar.dart [hunk 1]
```diff
@@ -0,0 +1,462 @@
...
+  });
+}
+
+extension on TextEditingController {
+  String get selectionText =>
+      text.substring(selection.baseOffset, selection.extentOffset);
+  String get beforeSelectionText => text.substring(0, selection.baseOffset);
+  String get afterSelectionText => text.substring(selection.extentOffset);
+
+  /// surroungs selection with given strings. If nothing is selected, placeholder is used in the middle
```
Comment (shilangyu) [inline lib/widgets/editor/editor_toolbar.dart:38]: Typo: surrounds

File: lib/widgets/editor/editor_toolbar.dart [hunk 2]
```diff
@@ -0,0 +1,462 @@
...
+      final finalString = '(${titleController.text})[$link]';
+      Navigator.of(context).pop(Reformat(
+        text: finalString,
+        selectionBeginningShift: finalString.length,
+        selectionEndingShift: finalString.length - selection.length,
+      ));
+    }
+
+    return AlertDialog(
+      title: const Text('Add link'),
```
Comment (shilangyu) [inline lib/widgets/editor/editor_toolbar.dart:426]: l10n

File: lib/widgets/editor/editor_toolbar.dart [hunk 3]
```diff
@@ -0,0 +1,462 @@
...
+        url = selection.startsWith(_websiteRegex) ? selection : '';
+
+  @override
+  Widget build(BuildContext context) {
+    final titleController = useTextEditingController(text: title);
+    final urlController = useTextEditingController(text: url);
+
+    void submit() {
+      final link = () {
+        if (urlController.text.startsWith('http?s://')) {
```
Comment (shilangyu) [inline lib/widgets/editor/editor_toolbar.dart:411]: typo `p?s`?

File: lib/widgets/editor/editor_toolbar.dart [hunk 4]
```diff
@@ -0,0 +1,462 @@
...
+          icon: const Icon(Icons.question_mark),
+          tooltip: L10n.of(context).editor_help,
+        ),
+      ],
+    );
+  }
+}
+
+class AddLinkDialog extends HookWidget {
+  final String title;
```
Comment (shilangyu) [inline lib/widgets/editor/editor_toolbar.dart:394]: This is called `text` or `label`, title is a different thing:
image

File: lib/widgets/editor/editor_toolbar.dart [hunk 5]
```diff
@@ -0,0 +1,462 @@
...
+
+    void submit() {
+      final link = () {
+        if (urlController.text.startsWith('http?s://')) {
+          return urlController.text;
+        } else {
+          return 'https://${urlController.text}';
+        }
+      }();
+      final finalString = '(${titleController.text})[$link]';
```
Comment (shilangyu) [inline lib/widgets/editor/editor_toolbar.dart:417]: Should be `'[${titleController.text}]($link)'`

File: lib/widgets/editor/editor_toolbar.dart [hunk 6]
```diff
@@ -0,0 +1,462 @@
...
+import '../../util/extensions/api.dart';
+import '../../util/extensions/spaced.dart';
+import '../../util/files.dart';
+import '../../util/mobx_provider.dart';
+import '../../util/observer_consumers.dart';
+import '../../util/text_lines_iterator.dart';
+import 'editor_picking_dialog.dart';
+import 'editor_toolbar_store.dart';
+
+class Reformat {
```
Comment (shilangyu) [inline lib/widgets/editor/editor_toolbar.dart:21]: Should be private probably

File: lib/widgets/editor/editor_toolbar.dart [hunk 7]
```diff
@@ -0,0 +1,462 @@
+import 'package:flutter/foundation.dart';
```
Comment (shilangyu) [inline lib/widgets/editor/editor_toolbar.dart:1]: (unrelated to this import) Personally, I would move helpers to the bottom of the file. Since EditorToolbar is the main focus

File: lib/widgets/editor/editor_toolbar.dart [hunk 8]
```diff
@@ -0,0 +1,462 @@
...
+    }
+
+    return AlertDialog(
+      title: const Text('Add link'),
+      content: Column(
+        mainAxisSize: MainAxisSize.min,
+        children: [
+          TextField(
+            controller: titleController,
+            decoration: const InputDecoration(hintText: 'title'),
```
Comment (shilangyu) [inline lib/widgets/editor/editor_toolbar.dart:432]: l10n

File: lib/widgets/editor/editor_toolbar.dart [hunk 9]
```diff
@@ -0,0 +1,462 @@
...
+            decoration: const InputDecoration(hintText: 'https://example.com'),
+            onEditingComplete: submit,
+            autocorrect: false,
+          ),
+        ].spaced(10),
+      ),
+      actions: [
+        TextButton(
+            onPressed: () => Navigator.of(context).pop(),
+            child: const Text('Cancel')),
```
Comment (shilangyu) [inline lib/widgets/editor/editor_toolbar.dart:447]: l10n

File: lib/widgets/editor/editor_toolbar.dart [hunk 10]
```diff
@@ -0,0 +1,462 @@
...
+        ElevatedButton(
+          onPressed: submit,
+          child: const Text('Add link'),
```
Comment (shilangyu) [inline lib/widgets/editor/editor_toolbar.dart:450]: l10n

File: lib/widgets/editor/editor_toolbar.dart [hunk 11]
```diff
@@ -0,0 +1,462 @@
...
+        IconButton(
+          onPressed: () => controller.surround(
+              before: '**',
+              placeholder: L10n.of(context).insert_text_here_placeholder),
+          icon: const Icon(Icons.format_bold),
+          tooltip: L10n.of(context).editor_bold,
+        ),
+        IconButton(
+          onPressed: () => controller.surround(
+              before: '*',
```
Comment (shilangyu) [inline lib/widgets/editor/editor_toolbar.dart:215]: `'_'` is more popular for italics

File: lib/widgets/editor/editor_toolbar.dart [hunk 12]
```diff
@@ -0,0 +1,462 @@
...
+                  return;
+                }
+                try {
+                  // FIXME: for some reason it doesn't go past this line on iOS. idk why
+                  final pic = await pickImage();
+                  // pic is null when the picker was cancelled
+
+                  if (pic != null) {
+                    final picUrl = await context
+                        .read<EditorToolbarStore>()
```
Comment (shilangyu) [inline lib/widgets/editor/editor_toolbar.dart:244]: You already have `store` in this scope

File: lib/widgets/editor/editor_toolbar.dart [hunk 13]
```diff
@@ -0,0 +1,462 @@
...
+        // Insert image
+        ObserverBuilder<EditorToolbarStore>(
+          builder: (context, store) {
+            return IconButton(
+              onPressed: loggedInAction((token) async {
+                if (store.imageUploadState.isLoading) {
+                  return;
+                }
+                try {
+                  // FIXME: for some reason it doesn't go past this line on iOS. idk why
```
Comment (shilangyu) [inline lib/widgets/editor/editor_toolbar.dart:238]: Is that for later?

File: lib/widgets/editor/editor_toolbar.dart [hunk 14]
```diff
@@ -0,0 +1,462 @@
...
+                        .read<EditorToolbarStore>()
+                        .uploadImage(pic.path, token);
+
+                    if (picUrl != null) {
+                      controller.reformatSimple('![]($picUrl)');
+                    }
+                  }
+                } on Exception catch (_) {
+                  ScaffoldMessenger.of(context).showSnackBar(
+                      const SnackBar(content: Text('Failed to upload image')));
```
Comment (shilangyu) [inline lib/widgets/editor/editor_toolbar.dart:253]: l10n

File: lib/widgets/editor/editor_toolbar.dart [hunk 15]
```diff
@@ -0,0 +1,462 @@
...
+          },
+          icon: const Icon(Icons.home),
+          tooltip: L10n.of(context).editor_community,
+        ),
+        PopupMenuButton<HeaderLevel>(
+          itemBuilder: (context) => [
+            for (final h in HeaderLevel.values)
+              PopupMenuItem(
+                value: h,
+                child: Text(describeEnum(h).toUpperCase()),
```
Comment (shilangyu) [inline lib/widgets/editor/editor_toolbar.dart:297]:
```suggestion
                child: Text(h.name.toUpperCase()),
```

File: lib/widgets/editor/editor_toolbar.dart [hunk 16]
```diff
@@ -0,0 +1,462 @@
...
+          onPressed: () {
+            final line = controller.firstSelectedLine;
+
+            if (line.startsWith(RegExp.escape('* '))) {
+              controller.removeAtBeginningOfEverySelectedLine('* ');
+            } else if (line.startsWith('- ')) {
+              controller.removeAtBeginningOfEverySelectedLine('- ');
+            } else {
+              controller.insertAtBeginningOfEverySelectedLine('- ');
+            }
```
Comment (shilangyu) [inline lib/widgets/editor/editor_toolbar.dart:335]: `+`

It would be nice if you kept all of these markdown markers in a centralized place. Then this code should handle a list of symbols instead of hardcoding some ifs

File: lib/widgets/editor/editor_toolbar.dart [hunk 17]
```diff
@@ -0,0 +1,462 @@
...
+          icon: const Icon(Icons.superscript),
+          tooltip: L10n.of(context).editor_superscript,
+        ),
+        //spoiler
+        IconButton(
+          onPressed: () {
+            controller.reformat((selection) {
+              final insides = selection.isNotEmpty ? selection : '___';
+              Logger.root
+                  .info([21, 21 + insides.length, insides, insides.length]);
```
Comment (shilangyu) [inline lib/widgets/editor/editor_toolbar.dart:370]: Debug?

File: lib/widgets/editor/editor_toolbar.dart [hunk 18]
```diff
@@ -0,0 +1,462 @@
...
+              return Reformat(
+                text: '\n::: spoiler spoiler\n$insides\n:::\n',
```
Comment (shilangyu) [inline lib/widgets/editor/editor_toolbar.dart:372]: l10n

File: lib/widgets/editor/editor_toolbar.dart [hunk 19]
```diff
@@ -0,0 +1,462 @@
...
+                selectionBeginningShift: 21,
+                selectionEndingShift: 21 + insides.length - selection.length,
```
Comment (shilangyu) [inline lib/widgets/editor/editor_toolbar.dart:374]: What is this magical number?

File: lib/widgets/editor/editor_toolbar.dart [hunk 20]
```diff
@@ -0,0 +1,462 @@
...
+
+  Toolbar({
+    required this.controller,
+    required this.instanceHost,
+  }) : store = EditorToolbarStore(instanceHost);
+
+  @override
+  Widget build(BuildContext context) {
+    return MobxProvider.value(
+      value: store,
```
Comment (shilangyu) [inline lib/widgets/editor/editor_toolbar.dart:168]: Please provide with the normal MobxProvider (not `.value`). Otherwise it is never disposed.

File: lib/widgets/editor/editor_toolbar.dart [hunk 21]
```diff
@@ -0,0 +1,462 @@
...
+  h3(3),
+  h4(4),
+  h5(5),
+  h6(6);
+
+  const HeaderLevel(this.value);
+  final int value;
+}
+
+class Toolbar extends HookWidget {
```
Comment (shilangyu) [inline lib/widgets/editor/editor_toolbar.dart:154]: Not a hook widget.

File: lib/widgets/editor/editor_toolbar.dart [hunk 22]
```diff
@@ -0,0 +1,462 @@
...
+      ),
+    );
+  }
+
+  void insertAtBeginningOfEverySelectedLine(String s) {
+    final lines = TextLinesIterator.fromController(this);
+    var linesCount = 0;
+    while (lines.moveNext()) {
+      if (lines.isWithinSelection) {
+        if (!lines.current.startsWith(RegExp.escape(s))) {
```
Comment (shilangyu) [inline lib/widgets/editor/editor_toolbar.dart:107]: Why RegExp.escape here?

File: lib/widgets/editor/editor_toolbar.dart [hunk 23]
```diff
@@ -0,0 +1,462 @@
...
+      ),
+    );
+  }
+
+  void removeAtBeginningOfEverySelectedLine(String s) {
+    final lines = TextLinesIterator.fromController(this);
+    var linesCount = 0;
+    while (lines.moveNext()) {
+      if (lines.isWithinSelection) {
+        if (lines.current.startsWith(RegExp.escape(s))) {
```
Comment (shilangyu) [inline lib/widgets/editor/editor_toolbar.dart:86]: ditto

File: lib/util/text_lines_iterator.dart [hunk 24]
```diff
@@ -0,0 +1,89 @@
...
+  int beg;
+  int end;
+  TextSelection? selection;
+
+  TextLinesIterator(this.text, {this.selection})
+      : end = -1,
+        beg = -1;
+
+  factory TextLinesIterator.fromController(TextEditingController controller) =>
+      TextLinesIterator(controller.text, selection: controller.selection);
```
Comment (shilangyu) [inline lib/util/text_lines_iterator.dart:15]:
```suggestion
  TextLinesIterator.fromController(TextEditingController controller) :
      this(controller.text, selection: controller.selection);
```

File: lib/util/text_lines_iterator.dart [hunk 25]
```diff
@@ -0,0 +1,89 @@
...
+
+  /// returns the lines as a list but also moves the pointer to the back
+  List<String> get asList {
+    reset();
+    final list = <String>[];
+    while (moveNext()) {
+      list.add(current);
+    }
+    return list;
+  }
```
Comment (shilangyu) [inline lib/util/text_lines_iterator.dart:88]: This is something an iterable should be doing, not an iterator

File: lib/util/text_lines_iterator.dart [hunk 26]
```diff
@@ -0,0 +1,89 @@
+import 'package:flutter/widgets.dart';
+
+/// utililty class for traversing through multiline text
+class TextLinesIterator extends Iterator {
```
Comment (shilangyu) [inline lib/util/text_lines_iterator.dart:4]: I have no idea what this iterator does and what are its elements. This public `set current` scares me

**Review by shilangyu — COMMENTED** [review]
Overall keeping strings as the source of truth is suboptimal. Keeping a custom data structure (preferably markdown AST) would make it more reliable, since currently you are doing many things manually (such as list detection, which handles only cases you thought of, but not all cases markdown supports (for example a list item does not necessarily have to span a single line)).

To keep the toolbar always above the keyboard it can be done the way you did. But it is not very neat since it is non-local and requires changing other parts of the page (which is directly an assumption that you have access to the full screen). To allow Editor itself to draw such a toolbar, you need to use `Overlay`. It is basically a Stack, but over your whole application and can be modified from anywhere. Sadly `Overlay`'s API is very clunky since it is imperative and requires manual mount/dismount of widgets. Thus instead, I would recommend [flutter_portal](https://pub.dev/packages/flutter_portal), which is a declarative implementation. But to be honest, I dont think this floating toolbar is always good. For example in manage_account, there are many text fields there, it is not obvious which textfield this floating toolbar would be related to. I would argue that even on create post the title/url textfields seem like they could be controlled by the toolbar, and yet they cannot. Maybe the toolbar should be shown only when there is focus on the editor? This would also enable the possibility of multiple Editors on a single page. Current implementation does not enable such functionality. 

Also, some tests would be nice

File: lib/widgets/editor/editor_toolbar.dart (hunk 22 above)
Comment (krawieck) [reply lib/widgets/editor/editor_toolbar.dart:107]: startswith parses the string as regex

File: lib/util/text_lines_iterator.dart (hunk 26 above)
Comment (krawieck) [reply lib/util/text_lines_iterator.dart:4]: no need to fear :-)

Comment (krawieck) [reply lib/util/text_lines_iterator.dart:4]: no need to fear :^)

File: lib/widgets/editor/editor_toolbar.dart (hunk 22 above)
Comment (shilangyu) [reply lib/widgets/editor/editor_toolbar.dart:107]: Are you sure? https://api.flutter.dev/flutter/dart-core/String/startsWith.html
It accepts a Pattern, where a normal string is a literal pattern

File: lib/formatter.dart [hunk 27]
```diff
@@ -0,0 +1,90 @@
...
+    return 0;
+  }
+
+  int getEndOfTheLine(int from) {
+    for (var i = from; i < length; i++) {
+      if (this[i] == '\n') return i;
+    }
+
+    return length - 1;
+  }
```
[the remark this answers is not in the document]
Comment (krawieck) [reply lib/formatter.dart:18]: `charactedIndex` doesn't convey any meaning. but i will switch `startingIndex` to `from`

File: lib/formatter.dart [hunk 28]
```diff
@@ -0,0 +1,90 @@
...
+        final indent = match.group(1);
+        final number = int.parse(match.group(2)!) + 1;
+
+        return tev.append('$indent$number$afterNumberChar ');
+      }
+
+      newVal = listContinuation('-', newVal);
+      newVal = listContinuation('*', newVal);
+      newVal = numberedListContinuation('.', newVal);
+      newVal = numberedListContinuation(')', newVal);
```
[the remark this answers is not in the document]
Comment (krawieck) [reply lib/formatter.dart:85]: yes, and only one can run at one pass

File: lib/widgets/editor/editor_picking_dialog.dart [hunk 29]
```diff
@@ -0,0 +1,132 @@
...
+            Navigator.of(context).pop(suggestion),
+        loadingBuilder: (context) => Row(
+          mainAxisAlignment: MainAxisAlignment.center,
+          children: const [
+            Padding(
+              padding: EdgeInsets.all(16),
+              child: CircularProgressIndicator.adaptive(),
+            ),
+          ],
+        ),
```
[the remark this answers is not in the document]
Comment (krawieck) [reply lib/widgets/editor/editor_picking_dialog.dart:55]: currently:
image

with center:
image

File: lib/widgets/editor/editor_picking_dialog.dart [hunk 30]
```diff
@@ -0,0 +1,132 @@
...
+        suggestionsCallback: (pattern) async {
+          if (pattern.trim().isEmpty) return const Iterable.empty();
+          return LemmyApiV3(store.instanceHost)
+              .run(Search(
+                q: pattern,
+                auth: userData?.jwt.raw,
+                type: SearchType.users,
+                limit: 10,
+              ))
+              .then((value) => value.users);
```
[the remark this answers is not in the document]
Comment (krawieck) [reply lib/widgets/editor/editor_picking_dialog.dart:34]: i don't think there was such agreement. small stuff is fine. other pickers also doesn't bother with a store

File: lib/widgets/editor/editor_toolbar.dart (hunk 22 above)
Comment (krawieck) [reply lib/widgets/editor/editor_toolbar.dart:107]: well now i see that i thought wrong for a very long time 😅

File: lib/widgets/editor/editor_toolbar.dart (hunk 18 above)
Comment (krawieck) [reply lib/widgets/editor/editor_toolbar.dart:372]: i don't think theres a need in this case

File: lib/widgets/editor/editor_toolbar.dart (hunk 19 above)
Comment (krawieck) [reply lib/widgets/editor/editor_toolbar.dart:374]: length of the string above '\n::: spoiler spoiler\n...\n:::\n'

File: lib/widgets/editor/editor_toolbar.dart (hunk 11 above)
Comment (krawieck) [reply lib/widgets/editor/editor_toolbar.dart:215]: no

File: lib/widgets/editor/editor_toolbar.dart (hunk 13 above)
Comment (krawieck) [reply lib/widgets/editor/editor_toolbar.dart:238]: it was a problem that was resolved at very late stage of this PR, can be removed now

File: lib/formatter.dart (hunk 27 above)
[the remark this answers is not in the document]
Comment (krawieck) [reply lib/formatter.dart:18]: i don't understand this sentence

File: lib/formatter.dart [hunk 31]
```diff
@@ -0,0 +1,90 @@
...
+/// Provides convinience formatting in markdown text fields
+class MarkdownFormatter extends TextInputFormatter {
+  @override
+  TextEditingValue formatEditUpdate(
+      TextEditingValue oldValue, TextEditingValue newValue) {
+    if (oldValue.text.length > newValue.text.length) return newValue;
+
+    var newVal = newValue;
+
+    final char = newValue.text[newValue.selection.baseOffset - 1];
```
[the remark this answers is not in the document]
Comment (krawieck) [reply lib/formatter.dart:51]: this doesn't happen cuz it only gets called when a char gets added. when a char gets added the cursor moves 1 forward so baseOffset can't be less than 1

File: lib/formatter.dart (hunk 27 above)
[the remark this answers is not in the document]
Comment (shilangyu) [reply lib/formatter.dart]: You are returning an inclusive index. Meaning, the returned index points to the last character in the line, not to one after that. Usually end indices are represented as exclusive ones. For instance, `'\nSome text'.substring(getBeginOfTheLine(2), getEndOfTheLine(2))` will return `'Some tex'` instead of `'Some text'`

File: lib/widgets/editor/editor_toolbar.dart (hunk 16 above)
Comment (shilangyu) [reply lib/widgets/editor/editor_toolbar.dart]: bump

File: lib/markdown_formatter.dart [hunk 32]
```diff
@@ -0,0 +1,92 @@
...
+        final number = int.tryParse(match.group(2)!) ?? 0 + 1;
+
+        return tev.append('$indent$number$afterNumberChar ');
+      }
+
+      newVal = unorderedListContinuation('-', newVal);
+      newVal = unorderedListContinuation('*', newVal);
+      newVal = unorderedListContinuation('+', newVal);
+      newVal = orderedListContinuation('.', newVal);
+      newVal = orderedListContinuation(')', newVal);
```
Comment (shilangyu) [inline lib/markdown_formatter.dart:87]: I think this should not create a new list item if the current one is empty. It should probably remove that empty list item and stop creating new items.

File: lib/widgets/editor/editor_toolbar.dart [hunk 33]
```diff
@@ -0,0 +1,469 @@
...
+    return MobxProvider(
+      create: (context) => store,
+      child: AsyncStoreListener(
+        asyncStore: store.imageUploadState,
+        child: visible.value
+            ? Container(
+                height: _height,
+                width: double.infinity,
+                color: Theme.of(context).cardColor,
+                child: Material(
```
Comment (shilangyu) [inline lib/widgets/editor/editor_toolbar.dart:75]: What does this Material do?

File: lib/widgets/editor/editor_toolbar.dart [hunk 34]
```diff
@@ -0,0 +1,469 @@
...
+                    ),
+                  ),
+                ),
+              )
+            : const SizedBox.shrink(),
+      ),
+    );
+  }
+
+  static Widget safeArea = const SizedBox(height: _height);
```
Comment (shilangyu) [inline lib/widgets/editor/editor_toolbar.dart:90]:
```suggestion
  static const safeArea = SizedBox(height: _height);
```

File: lib/widgets/editor/editor_toolbar.dart (hunk 16 above)
Comment (krawieck) [reply lib/widgets/editor/editor_toolbar.dart]: i'll see what i can do

File: lib/widgets/editor/editor_toolbar.dart (hunk 19 above)
Comment (shilangyu) [reply lib/widgets/editor/editor_toolbar.dart:270]: Then it should be computed off that string

File: lib/widgets/editor/editor_toolbar.dart [hunk 35]
```diff
@@ -0,0 +1,469 @@
...
+    useEffect(() {
+      editorFocusNode.addListener(() {
+        visible.value = editorFocusNode.hasFocus;
+      });
+
+      return;
+    }, [editorFocusNode]);
+
+    return MobxProvider(
+      create: (context) => store,
```
Comment (shilangyu) [inline lib/widgets/editor/editor_toolbar.dart:67]: Why like this? Why not created here? I understand one extra `Builder` might seem annoying, but it would be really beneficial to have a clear source of this store

File: lib/widgets/editor/editor_toolbar.dart [hunk 36]
```diff
@@ -0,0 +1,469 @@
...
+  @override
+  Widget build(BuildContext context) {
+    final visible = useState(editorFocusNode.hasFocus);
+    useEffect(() {
+      editorFocusNode.addListener(() {
+        visible.value = editorFocusNode.hasFocus;
+      });
+
+      return;
+    }, [editorFocusNode]);
```
Comment (shilangyu) [inline lib/widgets/editor/editor_toolbar.dart:64]:
```suggestion
    final visible = useListenable(editorFocusNode).hasFocus;
```

File: lib/widgets/write_comment.dart [hunk 37]
```diff
@@ -99,37 +99,54 @@ class WriteComment extends HookWidget {
...
+              mainAxisAlignment: MainAxisAlignment.spaceBetween,
+              children: [
+                const Spacer(),
+                Toolbar(
+                  controller: controller,
+                  instanceHost: post.instanceHost,
+                ),
+              ],
+            ),
+          ),
```
[the remark this answers is not in the document]
Comment (shilangyu) [reply lib/widgets/write_comment.dart]: Because you placed the SafeArea above Positioned (Positioned has to be the direct child of a Stack)

File: lib/widgets/editor/editor.dart [hunk 38]
```diff
@@ -49,17 +52,22 @@ class Editor extends HookWidget {
...
-      keyboardType: TextInputType.multiline,
-      textCapitalization: TextCapitalization.sentences,
-      onChanged: onChanged,
-      onSubmitted: onSubmitted,
-      maxLines: maxLines,
-      minLines: minLines,
-      decoration: InputDecoration(labelText: labelText),
+    return FocusScope(
+      child: Focus(
+        focusNode: focusNode,
```
Comment (shilangyu) [inline lib/widgets/editor/editor.dart:57]: Why is this needed?

**Review by shilangyu — COMMENTED** [review]
I **really** dislike how disruptive and non-local this EditorToolbar is (the diff it generated alone is sad). If you _really_ want to pursue this implementation then it could probably benefit from some HOC implementation.

File: lib/widgets/editor/editor_toolbar.dart (hunk 11 above)
Comment (krawieck) [reply lib/widgets/editor/editor_toolbar.dart:137]: lemmy-ui uses asterisks
image

File: lib/widgets/editor/editor_toolbar.dart (hunk 16 above)
Comment (krawieck) [reply lib/widgets/editor/editor_toolbar.dart]: done

File: lib/util/text_lines_iterator.dart (hunk 25 above)
Comment (krawieck) [reply lib/util/text_lines_iterator.dart:88]: it's just for debugging

File: lib/widgets/editor/editor_toolbar.dart [hunk 39]
```diff
@@ -0,0 +1,504 @@
...
+
+class BottomSticky extends StatelessWidget {
+  final Widget child;
+  const BottomSticky({required this.child});
+
+  @override
+  Widget build(BuildContext context) => Positioned(
+        bottom: 0,
+        child: SizedBox(
+          width: MediaQuery.of(context).size.width,
```
Comment (shilangyu) [inline lib/widgets/editor/editor_toolbar.dart:106]: `width: double.infinity` wont suffice?

**Review by shilangyu — APPROVED** [review]

File: lib/widgets/editor/editor_toolbar.dart (hunk 39 above)
Comment (krawieck) [reply lib/widgets/editor/editor_toolbar.dart:106]: it was causing an error, that was the reason why previously i was reluctant to use Positioned, because i couldn't figure out what was messing up
